File: zero-business/zero-main-business/zero-main-domain-business/src/main/scala/com/zjj/main/domain/role/RoleFactory.scala
package com.zjj.main.domain.role

import com.zjj.bean.componenet.BeanHelper
import com.zjj.i18n.component.BaseI18nException
import org.springframework.beans.BeanUtils
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils

/**
 * @author zengJiaJun
 * @crateTime 2024年11月21日 21:12
 * @version 1.0
 */
@Component
class RoleFactory(roleRepository: RoleRepository) {

  def create(cmd: StockInRoleCmd): Option[RoleAgg] = {
    if (cmd.id != null) {
      return roleRepository
        .findById(cmd.id)
        .peek(role => Option.when(StringUtils.hasText(cmd.roleKey) && !role.roleKey.equals(cmd.roleKey))(throw new BaseI18nException("roleKey.cannot.modified")))
        .peek(role => BeanUtils.copyProperties(cmd, role))
        .map(Option.apply(_))
        .getOrElseThrow(() => new BaseI18nException("role.not.exists", "role.not.exists", cmd.id))
    }
    roleRepository
      .findByRoleKey(cmd.roleKey)
      .peek(_ => throw new BaseI18nException("roleKey.exists", "roleKey.exists", cmd.roleKey))

    Option.apply(BeanHelper.copyToObject(cmd, classOf[RoleAgg]))
  }

}

File: zero-business/zero-main-business/zero-main-domain-business/src/main/java/com/zjj/main/domain/role/RoleFactory.java
package com.zjj.main.domain.role;

import com.zjj.bean.componenet.BeanHelper;
import com.zjj.i18n.component.BaseI18nException;
import com.zjj.main.domain.role.cmd.StockInRoleCmd;
import io.vavr.control.Option;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

/**
 * @author zengJiaJun
 * @version 1.0
 * @crateTime 2025年03月24日 16:11
 */
@Component
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class RoleFactory {

    private final RoleRepository roleRepository;

    public Option<RoleAgg> create(StockInRoleCmd cmd) {

        RoleAgg roleAgg = null;

        if (cmd.id() != null) {
            roleAgg = roleRepository
                    .findById(cmd.id())
                    .peek(role -> {
                        if (StringUtils.hasText(cmd.code()) && !role.getCode().equals(cmd.code())) {
                            throw new BaseI18nException("roleKey.cannot.modified");
                        }
                    })
                    .map(role -> BeanHelper.copyToObject(role, RoleAgg.class))
                    .getOrElseThrow(() -> new BaseI18nException("role.not.exists", "role.not.exists", cmd.id()));
        }

        if (roleAgg == null && roleRepository.existsByCode(cmd.code())) {
            throw new BaseI18nException("roleKey.exists", "roleKey.exists", cmd.code());
        }

        return Option.of(roleAgg).orElse(Option.of(BeanHelper.copyToObject(cmd, RoleAgg.class)));
    }
}
